Generals/Code/GameEngine/Source/Common/RTS/TunnelTracker.cpp

```diff
@@ -254,7 +254,13 @@ void TunnelTracker::destroyObject( Object *obj, void * )
 	// heal all the objects within the tunnel system using the iterateContained function
 void TunnelTracker::healObjects(Real frames)
 {
+#if !RETAIL_COMPATIBLE_CRC
+	if (m_lastHealFrame == TheGameLogic->getFrame())
+		return; // Only heal objects once per frame.
```

Comment:
I think the better fix here is to move

```cpp
		TunnelTracker *tunnelSystem = controllingPlayer->getTunnelSystem();
		if (tunnelSystem)
		{
			tunnelSystem->healObjects(modData->m_framesForFullHeal);
		}
```

out of `TunnelContain::update` to a place that is updated once a frame

Reply:
I agree and did consider moving it to the player update. But that seems like a bigger refactor and could have unforeseen complications, whereas this change is almost as safe as you can get.

Do you suggest investigating a refactor as part of this fix?

Reply:
I'd suggest being cleaner than retail is already enough to satisfy the clean implementation requirement. I will investigate how feasible to acquire the respective data.

Reply:
I have increased solution complexity by 1000%.

Reply:
This is fine. This would have to be done anyway if we were to optimize performance.